### PR TITLE
Block initial broadcast once verified

### LINKED ISSUE #7926: Rebroadcast blocks before importing them to reduce the hop latency.
@rphmeier suggested this as an optimization. Apparently we have this on Ethereum already, thanks to @tomusdrw.

### PR DESCRIPTION
**Pull request by rakanalh**
Resolves #7926 

This should trigger pre-import prior to executing the block, but after all consensus checks and validations have been made on the received block.
As discussed with @gnunicorn, converting `BlockImportNotification` should serve the purpose of being able to know whether the notification is about a pre-import block or for an imported block, which opens the possibility to have other components react to a pre-imported block if that serves a specific functionality. 
Currently, the network layer would react to such a notification but calling `announce_block` which would distribute the block info to connected peers.

### CODE REVIEW

File: client/service/src/client/client.rs
Comment (gnunicorn) [inline client/service/src/client/client.rs:1719]:
```suggestion
				warn!("Block pre-import announcement error:\n{:?}", e);
```

**Review by gnunicorn — APPROVED** [review]
LGTM

File: .gitlab-ci.yml
Comment (rakanalh) [inline .gitlab-ci.yml:106]: WIP... checking CI

**Comment by rakanalh** [comment]
Testing shows that the network layer rejects this because the block is unknown through the chain.

https://github.com/paritytech/substrate/blob/3c9b031e449a6249dde07e00066848e0ee481ddf/client/network/src/protocol.rs#L1090-L1100

File: client/service/src/client/client.rs
Comment (bkchr) [inline client/service/src/client/client.rs:1717]: Can you please elaborate this design decision?

Why do we once drag the block through the full system, send some pre import notification and than notify the sync again?

With this change it also means that we will announce the block before importing, when we have build the block our own.

Comment (bkchr) [reply client/service/src/client/client.rs:1717]: Why isn't the sync layer in control over this, as it notifies the block import anyway?

This would mean, request a header check, the header is checked, we re-announce the block and after that we continue to import the block.

Yes this will clearly require more changes, but will be less spaghetti code.

Comment (gnunicorn) [reply client/service/src/client/client.rs:1717]: Sure, I suggested to do it this way.

> Why do we once drag the block through the full system, send some pre import notification and than notify the sync again?

Because that is what was asked for. Specifically, [as the ticket stated](https://github.com/paritytech/substrate/issues/7926) to announce the block (over the network) once after initial checks are done and once after the actual import happened.

The instance that currently triggers that 'announcing' is this components, doing that through the channels. To provide this features for all consensus engines, we do that once here when they trigger the actual import (after they have done the light-weight checks) but before heavy lifting of actual execution happens.

> With this change it also means that we will announce the block before importing
That is the intention, yes. Please see the ticket.

> when we have build the block our own.
You correct, I missed that in my review, even though I said it before, This need to do a similar `make_notifications` check on `BlockOrigin` as execute does again later, but also filter out `Own`:
```rust
		let make_notifications = match import_block.origin {
			BlockOrigin::NetworkBroadcast |  BlockOrigin::ConsensusBroadcast => true,
			BlockOrigin::Genesis | BlockOrigin::NetworkInitialSync | BlockOrigin::Own | BlockOrigin::File => false,
		};
```
> Why isn't the sync layer in control over this, as it notifies the block import anyway?
> 
> This would mean, request a header check, the header is checked, we re-announce the block and after that we continue to import the block.

Not quite sure, what component you exactly mean here. To avoid confusion, could you point me to the part of that you are referring to?

The reason to put this here was because this isn't only checking the general signatures, but after the consensus's own `verify` has been run. So, not only is the 'header checked' to not be incorrect, the generic consensus check also happened, giving us more confidence that we are announcing a block we are probably not going to fail to import. If we announced without the consensus checks, we'd have a higher risk of announcing bad blocks - e.g. one check makes sure the block is actually not running against a finalized branch. It's a trade-off for sure but as the consensus checks are - compared to the actual state changes of an import - still pretty light weight, this is a decent optimization to comparable little risk.

That's the reasoning behind locating the pre-announcement here, in the 'consensus importing'-area, to have the validity of the general consensus checks and lower the risk of pre-announcing invalid blocks. Right now, there is no signal or other way to inform the outside that a block has reached that point. Hence the idea to extend the existing channel with the new signal.

An alternative approach we've considered would have been to create a yet another channel for that signal, but extending the existing channel with other types of signals and thus allow others to also react - e.g. this would allow for offchain-workers to be informed about a block pre-import - seemed more sensible.

**Comment by andresilva** [comment]
Regardless of the current design details, with the current implementation I am not entirely sure how this is supposed to work in practice. Let's look at an example of what will happen:

- Node1 starts importing block A 
- It does the initial verification checks
- It pre-announces the block
- It begins importing the block (for Polkadot we can assume this will take at most 2s but it could be more for other Substrate-based chains)
- In the meantime Node1 gets a request for block A, which it can't serve since it's not imported yet
- Node2 which requested block A from Node1 will reduce its reputation since it announced something that it couldn't serve

File: client/service/src/client/client.rs
Comment (bkchr) [reply client/service/src/client/client.rs:1717]: > Not quite sure, what component you exactly mean here. To avoid confusion, could you point me to the part of code that you are referring to?

Here we are notified about the block data: https://github.com/paritytech/substrate/blob/master/client/network/src/protocol.rs#L819

Than it will be processed here:
https://github.com/paritytech/substrate/blob/master/primitives/consensus/common/src/import_queue.rs#L176

And than it will be notified here again:
https://github.com/paritytech/substrate/blob/master/client/network/src/protocol.rs#L1293

Basically there is a lot of stuff that needs to be improved, especially in `import_queue.rs`, but this should be the main interaction point. We should communicate the result from there back to the requesting side, in our case the sync, that the verification was successful. 

That the sync layer is currently informed about what to announce over the service is also some old "artifact", which is mainly required for when WE build a block, but should not be required for when the sync layer requests to import something.

We should really try to prevent adding some new notification that is added for a subsystem that already has a direct "communication channel" with the block import. 


> e.g. this would allow for offchain-workers to be informed about a block pre-import - seemed more sensible.

Not sure what the benefit of this should be. Offchain-workers are executed on the state of the imported block, aka they can not do anything with this information that a block was verified correctly.

**Comment by gnunicorn** [comment]
pulling @rphmeier into this conversation, especially regarding the comment by @andresilva . Rob, thoughts?

**Comment by andresilva** [comment]
Currently the sync code always answers block requests by looking at what's on the database. I think for this feature to work the sync code will also need to be able to serve blocks that have been pre-imported but are still being imported, it will need to somehow manage a cache of what's in the `preimport -> import` stage. I don't have any concrete design suggestions for how to do this.

**Comment by rphmeier** [comment]
Yes, I agree with André - we need the sync code to be able to serve pre-verified blocks before they exist in the DB. And this overlay should be pruned after the blocks are in the DB or once verification fails.

@bkchr We'd like the checks that are done within `import_block` of GRANDPA and BABE to be done before broadcast. The main expense that we are trying to avoid is executing the block in the runtime and committing trie changes to RocksDB, AFAIK. This is difficult to do without a notification stream as the `BlockImport` traits are composed and ultimately form a synchronous wrapper around the client.

I see two ways forward if the extra notification stream in the client is to be avoided. I reserve judgement on whether it should be and defer to @gnunicorn and other maintainers on that decision. I do think that the addition of an extra notification stream is ugly.

1. Investigate the attack surface of doing only header-based checks and if acceptable, build the broadcast into import queue.
2. Re-design the `BlockImport` trait and interface to better support this use-case.

One way of redesigning might be to make `BlockImport` an `async trait`, where we introduce yet another layer of the onion that broadcasts pre-import notifications:

sketch:
```rust
struct BroadcastPreImportNotificationsBlockImport(Vec<Sender>, inner: BlockImport);

async fn import_block(...) {
    let notification = make_notification(...);
    for sender in senders { sender.send(notification).await } // no unbounded channel
    inner.import_block(...)
}
```

and this would be instantiated at the service level and hooked up to the sync protocol.

There are other, cleaner ways to accomplish the same thing but would require digging out even more internals

**Comment by bkchr** [comment]
> @bkchr We'd like the checks that are done within `import_block` of GRANDPA and BABE to be done before broadcast. When I had a brief look at these checks, especially those in GRANDPA, it seemed safer to have them run. The main expense that we are trying to avoid is executing the block in the runtime and committing trie changes to RocksDB, AFAIK. This is difficult to do without a notification stream as the `BlockImport` traits are composed and ultimately form a synchronous wrapper around the client.

Yeah, I did read and understood what you said. For me it is more how this communication is done here. We already have a communication between the import queue and sync. Sync instructs the import queue to import blocks and it gets notified by the task about the results. This is complelty independent from the rest of the system. 

I think it should stay this way. Adding some new notification type deep into substrate, is just some hack. We already have the problem that no one really knows when such a notification is send and it also changes from time to time. 

This import queue is currently implemented relative ugly, but could be improved, especially for what we want here. In ethereum we apparently only checked that the header was correct. In the end we should be able to do here the same. Even if all checks in the block import in grandpa and babe are correct, executing the block could still fail. Maybe we require some way in the sync protocol to invalidate a block announcement or something similar.

**Comment by rphmeier** [comment]
> In ethereum we apparently only checked that the header was correct. In the end we should be able to do here the same

Yes, however, it'll require some reorganization of GRANDPA code. Worth nothing that in Ethereum we only had PoW which has no on-disk metadata except the DAG which is just a function of block height. GRANDPA & BABE require a lot more on-disk metadata which we'd like to be written to the DB atomically with the trie changes.

Of course, these problems are all solvable, it just requires someone to roll up their sleeves, get a shovel, and dig @rakanalh  :)

File: client/service/src/client/client.rs
Comment (rakanalh) [reply client/service/src/client/client.rs:1717]: @bkchr As @gnunicorn mentioned, we opted to announce a pre-imported block right after the block was verified. in `import_single_block_metered` in `import_queue` is called by `import_many_blocks` after which `on_blocks_process` is called. At this point in `on_blocks_processed` all blocks have been imported which is a bit too late for the pre-import block rebroadcast to happen. We could have placed the pre-import notification in `import_single_block_metered` but @gnunicorn's opinion about this was that `Client::check_block` already contained the implementation for `finalized` and `import` notifications after executing the block so the best way for us was to announce pre-import there before the block was executed. As the discussion in the issue mentions, what we want to achieve is announcing the block **before** the block was executed as latency is incurred there.

> We should really try to prevent adding some new notification that is added for a subsystem that already has a direct "communication channel" with the block import.

I guess this is your biggest objection to the current implementation. I don't 100% understand what you mean by "direct communication channel with block import". But since `import_many_blocks` communicates `on_blocks_processed` back to the network through `BufferedLink`, we could atler the code here:
https://github.com/paritytech/substrate/blob/0b719f85222c85ff095585d128fac1979bbfdd38/client/network/src/service.rs#L1819-L1834
To add an additional method to the trait `on_block_verified` to the trait & the `impl` blocks so that this method once called, would handle the announcement of pre-imports... i am not 100% that would work but i am trying to find a middle ground here.
What do you think?

Comment (bkchr) [reply client/service/src/client/client.rs:1717]: > @bkchr As @gnunicorn mentioned, we opted to announce a pre-imported block right after the block was verified. in `import_single_block_metered` in `import_queue` is called by `import_many_blocks` after which `on_blocks_process` is called.

You know that you always only import one block when you sync at the tip of the chain? Maybe there are 2 when there is a fork, but it is very very unlikely that you download both forks at the same time.

> As the discussion in the issue mentions, what we want to achieve is announcing the block **before** the block was executed as latency is incurred there.

I can also repeat myself again, I have read the issue and understand the issue ;)



> I don't 100% understand what you mean by "direct communication channel with block import".

There is already a "channel" between the sync layer and the import queue. As already explained somewhere here, the following happens: Sync downloads a block, issues a block import, the block is imported and sync is informed about the imported block. What you want, is a step in between, that sends a notification back on the same "channel" as "block is imported".

The main problem here is that the current interfaces and implementation is shit in Substrate. We have this weird `NetworkLink` etc stuff that communicates with the `import-queue`.  

We could probably go the middle way you have proposed with the `on_block_verified` method. This is still better than adding a new announcement that is wired through the whole code base :+1:

**Comment by rakanalh** [comment]
@tomaka, @gnunicorn and I discussed recent changes and we'd like your opinion on this.

Following @bkchr's opinion on keeping the logic for announcing pre-import blocks isolated in the networking layer... the current implementation (still WIP) would differentiate between announcing pre-imported blocks from imported blocks but we end up sending the same message over the network. The differentiation happens internally since when announcing to peers, we would end up keeping the pre-imported headers locally so that any requests for a specific block (or header) would be answered from the list of pre-imported blocks maintained by the sync service. So for peers, it's going to be transparent whether the announcement is for a pre-imported block or an imported one.... that follows the suggestion made by @andresilva.

Any toughts on this approach?

**Comment by rakanalh** [comment]
> Currently the sync code always answers block requests by looking at what's on the database. I think for this feature to work the sync code will also need to be able to serve blocks that have been pre-imported but are still being imported, it will need to somehow manage a cache of what's in the `preimport -> import` stage. I don't have any concrete design suggestions for how to do this.

I have hit a roadblock regarding this. Apparently when some node (A) requests a block which was announced by Node (B), the block is not going to be registered in the backend which means that Node (A) will receive a response with 0 blocks. Therefore, disconnecting Node (B) and rejecting any further connections.

The problem with trying to resolve this issue is that the `BlockRequestHandler` only knows the client and can only serve blocks which are in the chain backend. However, knowledge about pre-imported blocks can exist only in the `Protocol`/`ChainSync` layers. 

The only immediate idea i have right now is to pull out the `BlockRequestHandler` logic which fetches the block(s) from the backend and somehow make it pass through the `ChainSync` instead. Which leaves us with:
`BlockImportHandler::handle_request` -> `ChainSync::block_by_id` -> `Either from client.header(block_id) or ChainSync::header(block_id)`.

Not sure how much effort that would take because the instance of `ChainSync` is burried under `NetworkWorker/Protocol/ChainSync`. The alternative is to pass down a channel to ChainSync to handle requests made by the `BlockRequestHandler`. 

Any other ideas?

File: client/api/src/client.rs
Comment (bkchr) [inline client/api/src/client.rs:114]: Why do we need this? 

You are again dragging sync specific information about the full nod, while this is all network specific.

Why don't you wire this just together with the `BlockRequestHandler`?

**Review by bkchr — CHANGES_REQUESTED** [review]

File: client/api/src/client.rs
Comment (rakanalh) [reply client/api/src/client.rs:114]: Think about the flow of data... we're verifying the block through `ChainSync` which doesn't have a direct link to `BlockRequestHandler` to communicate preimproted blocks. However, `ChainSync` does have an instance to the client to inform it about verified blocks. The `BlockRequestHandler` also asks the client for the block data so the client is the central to both components.

Comment (rakanalh) [reply client/api/src/client.rs:114]: Btw, i am not saying this is the best possible way... it's just that this is what can be achieved with the current architecture.

Comment (bkchr) [reply client/api/src/client.rs:114]: > we're verifying the block through `ChainSync` which doesn't have a direct link to `BlockRequestHandler` to communicate preimproted blocks

Just because there isn't a direct link yet, doesn't mean that it is the best solution to wire everything through the client. When designing systems you always try to have as much data locality as possible, which is clearly not the solution to move stuff through the client. 

If you check the code on how the network is setup, you see that the `BlockRequestHandler` is initialized before initializing the `NetworkWorker` which initializes the `Protocol` and this initializes the `ChainSync`. If you now would create some struct `VerifiedBlocks` or whatever that is passed to the `BlockRequestHandler` and to `ChainSync`, you would keep all this logic completely in the network layer.

Comment (rakanalh) [reply client/api/src/client.rs]: I implemented your suggestion

File: client/network/src/lib.rs
Comment (bkchr) [inline client/network/src/lib.rs:406]:
```suggestion
				self.blocks.read().iter().filter(|(_, (b, v))| b.header().number() == n && *v).map(|(_, (b, _))| b.extrinsics().to_vec())
```

File: client/network/src/lib.rs
Comment (bkchr) [inline client/network/src/lib.rs:358]: Why do we first try to check the client? Checking the preimport cache should be much faster.

File: client/network/src/lib.rs
Comment (bkchr) [inline client/network/src/lib.rs:365]:
```suggestion
		self.preimported_block_header(&*id).map(Ok).or_else(|| self.client.header(*id)).transpose()
```

Please make yourself familiar with combinators in rust

File: client/network/src/lib.rs
Comment (bkchr) [inline client/network/src/lib.rs:387]: Same as above.

File: client/network/src/lib.rs
Comment (bkchr) [inline client/network/src/lib.rs:343]:
```suggestion
	/// Create new instance of [`VerifiedBlocks`].
```

File: client/network/src/lib.rs
Comment (bkchr) [inline client/network/src/lib.rs:368]:
```suggestion
	fn pre_imported_block_header(&self, id: &BlockId<B>) -> Option<B::Header> {
```

And same changes to the function as below for the block body.

File: client/network/src/lib.rs
Comment (bkchr) [inline client/network/src/lib.rs:394]:
```suggestion
	fn pre_imported_block_body(&self, id: &BlockId<B>) -> Option<Vec<B::Extrinsic>> {
```

File: client/network/src/lib.rs
Comment (bkchr) [inline client/network/src/lib.rs:412]:
```suggestion
	fn block_verified(&self, id: &BlockId<B>) {
```

Comment (bkchr) [reply client/network/src/lib.rs:412]: And this should also only take the block hash.

File: client/network/src/protocol/sync.rs
Comment (bkchr) [inline client/network/src/protocol/sync.rs:1066]: Collect should be removed.

File: client/network/src/protocol/sync.rs
Comment (bkchr) [inline client/network/src/protocol/sync.rs:1061]:
```suggestion
		self.verified_blocks.extend(blocks.into_iter().filter_map(|b| {
```

clone makes no sense.

File: client/network/src/protocol/sync.rs
Comment (bkchr) [inline client/network/src/protocol/sync.rs:1060]: Missing docs.

File: client/network/src/protocol.rs
Comment (bkchr) [inline client/network/src/protocol.rs:1088]: If we don't announce this as best block, the other peers will not request this block. And no, the solution is not to pass `true` here.

You need to check the `BlockImportParams::ForkChoice` to check if the block is a new best block.

File: client/network/src/protocol.rs
Comment (bkchr) [inline client/network/src/protocol.rs:1118]: Why is this public? No docs?

File: client/network/src/protocol.rs
Comment (bkchr) [inline client/network/src/protocol.rs:1087]: Why is this public? Why no docs?

**Review by bkchr — CHANGES_REQUESTED** [review]
What is the status of this pr?

A lot of stuff is not documented, there is not one test added. 

I added quite a lot of comments.

I expect at least one test added to `sc-network-test` that uses some sort of "forgetting" backend to ensure that we propagate the blocks when we only have preimported them.

File: client/network/src/lib.rs
Comment (rakanalh) [reply client/network/src/lib.rs]: That's pretty nice! thanks for the feedback.

File: client/network/src/protocol.rs
Comment (rakanalh) [reply client/network/src/protocol.rs]: The announcement for an imported block on the networking layer looks at:
```
let is_best = self.chain.info().best_hash == header.hash();
```
To figure out whether a block is best or not. At pre-import, how can `BlockImportParams::ForkChoice` be helpful if we haven't imported the block yet?

File: client/network/src/lib.rs
Comment (rakanalh) [reply client/network/src/lib.rs:341]: Added some pruning logic after the block was imported.

**Pull request closed without merging**
